File: parse_data.py
import csv
import json

def trim(s):
    return s.strip() if s else ""

csv_path = 'datos.csv'
json_path = 'data.json'

records = []
current_region = ""

with open(csv_path, 'r', encoding='utf-8') as f:
    reader = csv.reader(f)
    header = next(reader)
    
    for row in reader:
        # Pad row to 5 elements just in case
        while len(row) < 5:
            row.append("")
            
        ciudad = trim(row[0])
        taller = trim(row[1])
        contacto = trim(row[2])
        marcas = trim(row[3])
        gps = trim(row[4])

        if ciudad and not taller and not contacto and not marcas and not gps:
            current_region = ciudad
            continue

        final_region = ciudad
        if not ciudad and (taller or marcas):
            final_region = current_region
        elif ciudad:
            current_region = ciudad

        if taller:
            lat, lng = None, None
            if gps:
                parts = gps.replace('"', '').split(',')
                if len(parts) == 2:
                    try:
                        lat = float(parts[0].strip())
                        lng = float(parts[1].strip())
                    except ValueError:
                        pass
            
            records.append({
                "ciudad": final_region.upper(),
                "nombre": taller,
                "contacto": contacto,
                "marcas": marcas,
                "lat": lat,
                "lng": lng,
                "gps": gps
            })

with open(json_path, 'w', encoding='utf-8') as f:
    json.dump(records, f, indent=4, ensure_ascii=False)

print(f"Successfully parsed {len(records)} workshops to data.json")

File: parse_data.js
const fs = require('fs');

const csvPath = 'datos.csv';
const jsonPath = 'data.json';

try {
    const csvContent = fs.readFileSync(csvPath, 'utf8');
    const lines = csvContent.split(/\r?\n/);
    const records = [];

    let currentRegion = '';

    // Index 0 is header: CIUDAD,TALLER,CONTACTO,MARCA,UBICACIÓN POR GPS
    for (let i = 1; i < lines.length; i++) {
        let line = lines[i].trim();
        if (!line) continue;

        // Parse CSV line handling quotes
        const row = [];
        let inQuote = false;
        let currentValue = '';

        for (let char of line) {
            if (char === '"') {
                inQuote = !inQuote;
            } else if (char === ',' && !inQuote) {
                row.push(currentValue);
                currentValue = '';
            } else {
                currentValue += char;
            }
        }
        row.push(currentValue);

        // Map columns
        const ciudad = row[0]?.trim() || '';
        const taller = row[1]?.trim() || '';
        const contacto = row[2]?.trim() || '';
        const marcas = row[3]?.trim() || '';
        const gps = row[4]?.trim() || '';

        // Detect region headers (e.g., "Sucre,,,,")
        if (ciudad && !taller && !contacto && !marcas && !gps) {
            currentRegion = ciudad;
            continue;
        }

        // Handle rows that belong to a region but might have empty 'ciudad'
        let finalRegion = ciudad;
        if (!ciudad && (taller || marcas)) {
            finalRegion = currentRegion;
        } else if (ciudad) {
            currentRegion = ciudad; // Update latest region
        }

        // Only add valid workshops
        if (taller) {
            // Clean up GPS coordinates if they exist
            let lat = null, lng = null;
            if (gps) {
                const parts = gps.split(',');
                if (parts.length === 2) {
                    lat = parseFloat(parts[0].trim());
                    lng = parseFloat(parts[1].trim());
                }
            }

            records.push({
                ciudad: finalRegion.toUpperCase(),
                nombre: taller,
                contacto: contacto,
                marcas: marcas,
                lat: lat,
                lng: lng,
                gps_original: gps
            });
        }
    }

    fs.writeFileSync(jsonPath, JSON.stringify(records, null, 2), 'utf8');
    console.log(`Successfully parsed ${records.length} workshops to data.json`);

} catch (e) {
    console.error('Error parsing CSV:', e);
}
